platform/linux-generic/pktio/mdev/i40e.c

```diff
@@ -0,0 +1,641 @@
+		     odp_packet_t pkt_table[], int num)
+{
+	pktio_ops_i40e_data_t *pkt_i40e = pktio_entry->s.ops_data;
+	i40e_rx_queue_t *rxq = &pkt_i40e->rx_queues[rxq_idx];
+	uint16_t refill_from;
+	int rx_pkts = 0;
+
+	if (!pkt_i40e->lockless_rx)
+		odp_ticketlock_lock(&rxq->lock);
+
+	/* Keep track of the start point to refill RX queue */
+	refill_from = rxq->cidx ? rxq->cidx - 1 : rxq->rx_queue_len - 1;
+
+	while (rx_pkts < num) {
+		volatile i40e_rx_desc_t *rxd = &rxq->rx_descs[rxq->cidx];
+		odp_packet_hdr_t *pkt_hdr;
+		odp_packet_t pkt;
+		uint16_t pkt_len;
+		uint64_t status_error_len;
+		int ret;
+
+		status_error_len = odp_le_to_cpu_64(rxd->status_error_len);
+		if (!status_error_len)
+			break;
+
+		pkt_len = (status_error_len & I40E_RXD_QW1_LEN_M) >>
+		    I40E_RXD_QW1_LEN_S;
+
+		pkt = odp_packet_alloc(pkt_i40e->pool, pkt_len);
+		if (odp_unlikely(pkt == ODP_PACKET_INVALID))
+			break;
+
+		pkt_hdr = odp_packet_hdr(pkt);
+
+		ret = odp_packet_copy_from_mem(pkt, 0, pkt_len,
+					       (uint8_t *)rxq->rx_data.vaddr +
+					       rxq->cidx * I40E_RX_BUF_SIZE);
+		if (odp_unlikely(ret)) {
+			odp_packet_free(pkt);
+			break;
+		}
+
+		pkt_hdr->input = pktio_entry->s.handle;
+
+		rxq->cidx++;
+		if (odp_unlikely(rxq->cidx >= rxq->rx_queue_len))
+			rxq->cidx = 0;
+
+		pkt_table[rx_pkts] = pkt;
+		rx_pkts++;
+	}
+
+	if (rx_pkts)
+		i40e_rx_refill(rxq, refill_from, rx_pkts);
+
+	if (!pkt_i40e->lockless_rx)
+		odp_ticketlock_unlock(&rxq->lock);
+
+	return rx_pkts;
+}
+
+/**
+ * Helper function to build descriptor 2nd quad-word.
+ */
+static inline uint64_t txd_ctol(uint32_t cmd, uint32_t tag, uint32_t offset,
+				uint32_t len)
+{
+	return I40E_TXD_DTYPE_DATA |
+	    ((uint64_t)cmd << I40E_TXD_QW1_CMD_S) |
+	    ((uint64_t)tag << I40E_TXD_QW1_L2TAG1_S) |
+	    ((uint64_t)offset << I40E_TXD_QW1_OFFSET_S) |
+	    ((uint64_t)len << I40E_TXD_QW1_LEN_S);
+}
+
+static int i40e_send(pktio_entry_t *pktio_entry, int txq_idx,
+		     const odp_packet_t pkt_table[], int num)
+{
+	pktio_ops_i40e_data_t *pkt_i40e = pktio_entry->s.ops_data;
+	i40e_tx_queue_t *txq = &pkt_i40e->tx_queues[txq_idx];
+	uint16_t budget, tx_txds = 0;
+	int tx_pkts = 0;
+
+	if (!pkt_i40e->lockless_tx)
+		odp_ticketlock_lock(&txq->lock);
+
+	/* Determine how many packets will fit in TX queue */
+	budget = txq->tx_queue_len - 1;
+	budget -= txq->pidx;
+	budget += odp_le_to_cpu_32(*txq->cidx);
+	budget &= txq->tx_queue_len - 1;
+
+	while (tx_txds < budget && tx_pkts < num) {
+		uint16_t pkt_len = _odp_packet_len(pkt_table[tx_pkts]);
+		uint32_t offset = txq->pidx * I40E_TX_BUF_SIZE;
+
+		i40e_tx_desc_t *txd = &txq->tx_descs[txq->pidx];
+
+		uint32_t txd_cmd = I40E_TXD_CMD_ICRC | I40E_TXD_CMD_EOP;
+
+		/* Skip undersized packets silently */
+		if (odp_unlikely(pkt_len < I40E_TX_PACKET_LEN_MIN)) {
+			tx_pkts++;
+			continue;
+		}
+
+		/* Skip oversized packets silently */
+		if (odp_unlikely(pkt_len > I40E_TX_BUF_SIZE)) {
+			tx_pkts++;
+			continue;
+		}
```

Comment:
Do you plan to have different behavior in the future for undersized vs. oversized packets? If not, it would seem better to combine these into a single `if` test.

Reply:
We'll need to add statistics counters to all these drivers at some point, so I thought that it's better to keep 'oversized' and 'undersized' separately with an intention to update different counters. If we find out that we don't have separate counters (I guess stats are covered by some RFC, isn't it ?) -- then we definitely can merge those into a single condition check.